From b67e95f9952c4993486f3d601e5ab2cc3f9efccc Mon Sep 17 00:00:00 2001
From: Myy <myy@miouyouyou.fr>
Date: Tue, 18 Jul 2017 23:40:07 +0000
Subject: [PATCH] Ugly MMC hack, trying to resolve Tinkerboard's reboot issue

Signed-off-by: Myy <myy@miouyouyou.fr>
---
 drivers/mmc/host/dw_mmc-rockchip.c | 52 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 52 insertions(+)

diff --git a/drivers/mmc/host/dw_mmc-rockchip.c b/drivers/mmc/host/dw_mmc-rockchip.c
index a3f1c2b..57ffb6a 100644
--- a/drivers/mmc/host/dw_mmc-rockchip.c
+++ b/drivers/mmc/host/dw_mmc-rockchip.c
@@ -19,6 +19,11 @@
 #include "dw_mmc.h"
 #include "dw_mmc-pltfm.h"
 
+// For the hack
+#include <linux/regulator/consumer.h>
+#include <linux/string.h>
+#include "../core/core.h"
+
 #define RK3288_CLKGEN_DIV       2
 
 struct dw_mci_rockchip_priv_data {
@@ -392,6 +397,53 @@ static struct platform_driver dw_mci_rockchip_pltfm_driver = {
 	},
 };
 
+/* A weird reboot hack
+ * Discussed here :
+ * https://github.com/Miouyouyou/MyyQi/issues/8
+ */
+static void dw_mci_rockchip_platfm_shutdown(struct platform_device *pdev)
+{
+	struct dw_mci *host = platform_get_drvdata(pdev);
+	struct mmc_host *mmc = host->slot->mmc;
+
+	int ret;
+	
+	printk(
+		KERN_ERR "( Myy ) Powering the MMC hardware OFF ! (Hack)\n"
+	);
+	mmc_power_off(mmc);
+	mdelay(20);
+	
+	if (!IS_ERR(mmc->supply.vmmc)) {
+		ret = regulator_enable(mmc->supply.vmmc);
+		printk(
+			KERN_ERR "( Myy ) Re-enabling the MMC regulator -> %d\n", ret
+		);
+	}
+	
+	if (!IS_ERR(mmc->supply.vqmmc)) {
+		ret = regulator_set_voltage(mmc->supply.vqmmc, 3000000, 3300000);
+		printk(KERN_ERR "( Myy ) Modified the MMC Voltage -> %d\n", ret);
+	}
+}
+
+static int dw_mci_rockchip_setup_tinker_reboot_fix(char *str)
+{
+	if (strstr(str, "on")) {
+		printk(KERN_ERR "( Myy ) Enabling Tinkerboard's reboot fix !\n");
+		dw_mci_rockchip_pltfm_driver.shutdown = 
+			dw_mci_rockchip_platfm_shutdown;
+	}
+	else 
+		printk(
+			KERN_INFO "( Myy ) Not enabling Tinkerboard's reboot fix\n"
+		);
+	
+	return 0;
+}
+
+__setup("rockchip_tinkerrebootfix=", dw_mci_rockchip_setup_tinker_reboot_fix);
+
 module_platform_driver(dw_mci_rockchip_pltfm_driver);
 
 MODULE_AUTHOR("Addy Ke <addy.ke@rock-chips.com>");
-- 
2.10.2

